### PR TITLE
Adding more CLI Tests

### PR DESCRIPTION
**Pull request by dread-pirate-johnny-spaceboots**
Added a few more CLI tests while trying to adhere to your comment structure. Had to pull the list of options back out into a class param as it's used in another test as well. Also moved the console stream overrides to a setup method and installed a seam around the file system access in Main.

I see now what you mean by 'C#y' I think, looks to boil down mostly to code formatting. In most projects I work on with other people I like to set an auto-linter as part of the CI process so that stuff gets handled automatically. Helps avoid back an forth in PRs and merge conflicts from simple formatting changes too.

I stopped at the first test wherein I had to made a structural decision, which is this: When testing we need to set up the state of the application to control the codepath it'll follow on execution, often this is where mocks come in. In the case of this particular project though the thing you really need to mock is the program data fed into the Z80 and there's a couple ways you could do that.
1. Create files and have them read in by the file system abstraction
2. Mock the filesystem abstraction and supply the program data as byte arrays defined in the tests
3. Mock the filesystem abstraction to return dummy data and create partial mocks of the Z80 to bootstrap its state in tests

In this PR I've implemented #2, IMO keeping the tests with their data (vs splitting them into many files ala #1) would be preferable. Plus you could potentially supply the bytecode in asm and run it through a C# Z80 assembler.

I had tooled about with mocking the Z80 emulator you're using here, but due to the way its implemented directly wiring its state in a simple mock isn't possible. The good news there is you don't really want to do that for this project anyway, so kudos to static typing for making dumb things hard.

### CODE REVIEW

**Comment by johnny-spaceboots** [comment]
Also, do you have a particular branching model/strategy you'd like to employ here? 

Feature branching is very popular, but I prefer TBD with proper testing myself, but will abide by anything.

File: Z80Bench.cs [hunk 1]
```diff
@@ -6,12 +6,9 @@
 
 namespace z80bench
 {
-    public static class Z80Bench
-    {
-        public static int Main(string[] args)
-        {
-            if (args.Length < 1)
-            {
+    public static class Z80Bench {
```
Comment (maxim-zhao) [inline Z80Bench.cs:9]: You seem to have inadvertently K&R'd all the curly braces 😱

**Comment by maxim-zhao** [comment]
I think it'd be cleaner to add an intermediate state between CLI parsing and the runner itself, in the form of a class holding all the test info - list of data/offset tuples for memory insertion, values for other options and comparisons. Then you can test the conversion from CLI to this state, and the result of running with a given test state, and avoid the need to inject CLI for non-CLI tests.

The re-static-init of the main class is only needed because of the VDP construction, which should probably be changed to instance scope anyway.

I don't really mind about feature branches vs. TBD for this as in GitHub your trunk is just another branch from my point of view anyway.

File: Z80Bench.cs (hunk 1 above)
Comment (johnny-spaceboots) [reply Z80Bench.cs:9]: 🤦 guess I pushed this accidently. Strange that my rebasing against your comment commit left it in that state though... although maybe I had some small local difference.

**Comment by johnny-spaceboots** [comment]
> I think it'd be cleaner to add an intermediate state between CLI parsing and the runner itself, in the form of a class holding all the test info - list of data/offset tuples for memory insertion, values for other options and comparisons. 

Are you referring to something like a data provider? ie, you setup multiple sets of data to be fed into a single test (which can have conditional assertions based on input).

> Then you can test the conversion from CLI to this state, and the result of running with a given test state, and avoid the need to inject CLI for non-CLI tests.

I'm not sure I'm quite following here in regard to the 'inject CLI for non-CLI tests'. By inject CLI are you referring to the seam around the filesystem access?

**Comment by maxim-zhao** [comment]
I'm part-way through it now :) Basically:

1. Make a (POCO) class that encapsulates all the inputs to the benchmark
2. Make a class that executes the benchmark and stores the result state
3. Make the CLI drive this by building (1) and passing it to an instance of (2), then print the results

This way we can test the parts more independently. To create a test case, you make the (1) and (2) objects and check the results; you don't need to create commandlines or parse the output text - except when you want to actually test that part. I'll push some changes to demonstrate...

**Comment by maxim-zhao** [comment]
Pushed some changes now, causing much conflict here of course. Hopefully it makes more sense?

**Comment by johnny-spaceboots** [comment]
> Pushed some changes now, causing much conflict here of course. Hopefully it makes more sense?

Yes, I get you now. Restructuring the program to be more testable was certainly the right approach. I'm just a little puzzled about this comment:

> // These cannot be added to...

From the Benchmark constructor, not sure what about the values it refers to (Memory, RamComparisons, VramComparisons) can't be added to, as they're declared public and also those would be the main points of interest for bootstrapping test state no?

**Comment by maxim-zhao** [comment]
> From the Benchmark constructor, not sure what about the values it refers to (Memory, RamComparisons, VramComparisons) can't be added to, as they're declared public and also those would be the main points of interest for bootstrapping test state no?

They are exposed as `IEnumerable` and as such it is impossible to add items to the collections from the outside (except by casting of course), only to replace the collections for each (which is entirely the intention for testing). Thus it's safe to just let them all point at the same empty collection. I could leave them as `null` but I prefer not to leave things in that state.

**Pull request closed without merging**